pkg/tenantctl/controller.go

```diff
@@ -157,219 +172,470 @@ func (c *Controller) processNextWorkItem() bool {
+	return nsName, err
+}
+
+func (c *Controller) rollbackTenantNamespace(namespace string, tenant *api.Tenant, config clientConfig) error {
+	if namespace == "" {
+		return nil
 	}
+	c.logPrintf(tenant, "rolling back tenant namespace %q", namespace)
+	namespaceManager := c.getNamespaceManager(config)
+	err := namespaceManager.Delete(namespace)
+	if err != nil {
+		err = errors.WithMessagef(err, "failed to rollback tenant namespace %q", namespace)
+		c.logPrintln(tenant, err)
+		return err
+	}
+	return nil
 }
 
-func (c *Controller) rollback(tenant *api.Tenant) error {
-	log.Printf("Rollback tenant %s", tenant.GetName())
-	if tenant.Status.TenantNamespaceName == "" {
-		log.Printf("Nothing to rollback for tenant %s", tenant.GetName())
-	} else {
-		err := c.deleteNamespace(tenant)
+func (c *Controller) syncTenantRoleBinding(tenant *api.Tenant, namespace string, config clientConfig) (bool, error) {
+	if c.testing != nil && c.testing.syncTenantRoleBindingStub != nil {
+		return c.testing.syncTenantRoleBindingStub(tenant, namespace, config)
+	}
+
+	/*
+		The roleRef of an existing RoleBinding cannot be updated (prohibited by
+		server). This means we need to create a new RoleBinding when we want to
+		change the tenant role.
+		We cannot simply delete the existing role and create a new one (in that
+		order), because that would revoke permissions for some time and may fail
+		API calls concurrently performed by the respective subjects.
+		Instead we will first create the new RoleBinding and remove the old one
+		afterwards.
+		To deal with remainders from preceding failed attempts, we expect that
+		an arbitrary number of RoleBinding objects may exist. Recreation takes
+		place if the number of existing role bindings is not exactly one, or the
+		single existing role binding is not up-to-date.
+		We manage only those RoleBinding objects that are marked as "managed by
+		Steward". All others will not be touched or taken into account.
+	*/
+
+	syncNeeded := false
+
+	err := func() error {
+		rbList, err := c.listManagedRoleBindings(namespace)
 		if err != nil {
-			log.Printf("ERROR: Deletion of %s failed: %v", tenant.Status.TenantNamespaceName, err.Error())
 			return err
 		}
+
+		clientNamespace := tenant.GetNamespace()
+		expectedTenantRB := c.generateTenantRoleBinding(namespace, clientNamespace, config)
+
+		if len(rbList.Items) != 1 || !c.isTenantRoleBindingUpToDate(&rbList.Items[0], expectedTenantRB) {
+			syncNeeded = true
+		}
+
+		if syncNeeded {
+			c.logPrintf(tenant, "updating RoleBinding in tenant namespace %q", namespace)
+			_, err = c.createRoleBinding(expectedTenantRB)
+			if err != nil {
+				return err
+			}
+			err = c.deleteRoleBindingsFromList(rbList)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}()
+
+	if err != nil {
+		err = errors.WithMessagef(err,
+			"failed to sync the RoleBinding in tenant namespace %q",
+			namespace,
+		)
+		c.logPrintln(tenant, err)
 	}
-	return nil
+	return syncNeeded, err
```

Comment:
Shouldn't we set `syncNeeded` to false if `err` is nil? Sync is not needed anymore.

Reply:
See method doc in added by 72e8a0b14d98b755f66e83c3ee9748ddca7368f1.